main.cpp

```diff
@@ -14,13 +18,19 @@ struct User {
     std::string phone;
 };
 
+auto now = std::chrono::steady_clock::now();
 std::map<std::string, User> users;
-std::map<std::string, long> has_login;  // 换成 std::chrono::seconds 之类的
+std::map<std::string, decltype(now)> has_login;  // 换成 std::chrono::seconds 之类的
+
+// user资源的读写锁
+std::shared_mutex mtx;
+
 
 // 作业要求1：把这些函数变成多线程安全的
 // 提示：能正确利用 shared_mutex 加分，用 lock_guard 系列加分
-bool do_register(std::string username, std::string password, std::string school, std::string phone) {
+std::string do_register(std::string username, std::string password, std::string school, std::string phone) {
     User user = {password, school, phone};
+    std::shared_lock grd{ mtx };
```

Comment:
```suggestion
    std::unique_lock grd{ mtx };
```
register是写入操作，需要独占，是unique_lock。

Reply:
啊。。读写全搞反了 -.-|||